langchain/llms/databricks.py

```diff
@@ -0,0 +1,323 @@
+import os
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Dict, List, Optional
+
+import requests
+from pydantic import BaseModel, Extra, Field, PrivateAttr, root_validator, validator
+
+from langchain.callbacks.manager import CallbackManagerForLLMRun
+from langchain.llms.base import LLM
+
+__all__ = ["Databricks"]
+
+
+class _DatabricksClientBase(BaseModel, ABC):
+    """A base JSON API client that talks to Databricks."""
+
+    api_url: str
+    api_token: str
+
+    def post_raw(self, request: Any) -> Any:
+        headers = {"Authorization": f"Bearer {self.api_token}"}
+        response = requests.post(self.api_url, headers=headers, json=request)
+        # TODO: error handling and automatic retries
+        if not response.ok:
+            raise ValueError(f"HTTP {response.status_code} error: {response.text}")
+        return response.json()
+
+    @abstractmethod
+    def post(self, request: Any) -> Any:
+        ...
+
+
+class _DatabricksServingEndpointClient(_DatabricksClientBase):
+    """An API client that talks to a Databricks serving endpoint."""
+
+    host: str
+    endpoint_name: str
+
+    @root_validator(pre=True)
+    def set_api_url(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        if "api_url" not in values:
+            host = values["host"]
+            endpoint_name = values["endpoint_name"]
+            api_url = f"https://{host}/serving-endpoints/{endpoint_name}/invocations"
+            values["api_url"] = api_url
+        return values
+
+    def post(self, request: Any) -> Any:
+        # See https://docs.databricks.com/machine-learning/model-serving/score-model-serving-endpoints.html
+        wrapped_request = {"dataframe_records": [request]}
+        response = self.post_raw(wrapped_request)["predictions"]
+        # For a single-record query, the result is not a list.
+        if isinstance(response, list):
+            response = response[0]
+        return response
+
+
+class _DatabricksClusterDriverProxyClient(_DatabricksClientBase):
+    """An API client that talks to a Databricks cluster driver proxy app."""
+
+    host: str
+    cluster_id: str
+    cluster_driver_port: str
+
+    @root_validator(pre=True)
+    def set_api_url(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        if "api_url" not in values:
+            host = values["host"]
+            cluster_id = values["cluster_id"]
+            port = values["cluster_driver_port"]
+            api_url = f"https://{host}/driver-proxy-api/o/0/{cluster_id}/{port}"
```

Comment:
@mengxr  This might have a problem with the databricks driver proxy api with /o/0. The databricks repl context provided apiUrl is something like http://oregon.staging.databricks.com/. From my set up it seems that this does not work with the /o/0.

Reply:
This works fine with databricks notebook context's url address. Resolve now.